File: solutions/javascript/two-bucket/2/two-bucket.js
//
// This is only a SKELETON file for the 'Two Bucket' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

const gcd = (a,b) => (b? gcd(b, a%b) : a);

export class TwoBucket {
  constructor(buckOne, buckTwo, goal, starterBuck) {
    if(goal > Math.max(buckOne, buckTwo))
      throw new Error("Not possible");  
    else if(goal % gcd(buckOne, buckTwo) != 0)
       throw new Error("Not possible");
     this.firstB_ = buckOne;
     this.secondB_ =  buckTwo;
     this.goal_ = goal;
     this.startB_ = starterBuck;
    }
  

  extendedGcd(a, b) {
    if(b==0){
      console.log("%d!\n", a);
      return [a,1,0];}
    var [d,u1,v1] = this.extendedGcd(b, a%b);
    console.log("a,b,d: %d, %d, %d\n", a,b, d);
    return [d, v1, u1 - v1*Math.floor(a/b)];
  }
  
   solveOne(one, two) {
     if(this.goal_ == one)
       return {moves: 1,  goalBucket: 'one', otherBucket: 0};
     else if(this.goal_ == two)
       return {moves: 2, goalBucket:'two', otherBucket: one};
     
    var [d,u,v] = this.extendedGcd(one, two);
    /*if (this.goal_ % d !==0) {   
          throw new Error('Not possible');
    } */
    u *= this.goal_ / d;
    v *= this.goal_ / d;
    
     if(u<0) {
      var t =  Math.floor(-u*d/two) + 1;
      u += t*(two/d);
      v -= t*(one/d);
    }
    if (one >= this.goal_)
      return {
        moves: Math.abs(2*(u-v) -2),
        goalBucket: 'one',
        otherBucket: two,
      }; 
    else
      return {
        moves: Math.abs(2*(u-v)),
        goalBucket: 'two',
        otherBucket: 0,
    }; 
  }
 
  solve() {
    if (this.goal_==0)
         return {moves: 0, goalBucket: this.startB_, otherBucket: 0};
      
    if (this.startB_ == 'one'){
        return this.solveOne(this.firstB_, this.secondB_); 
    }
    else {
      const res = this.solveOne(this.secondB_, this.firstB_);
      return {
        moves: res.moves,
        goalBucket: res.goalBucket == 'one' ? 'two' : 'one',
        otherBucket: res.otherBucket,
        };
      }
    }

  

}

File: solutions/dart/two-bucket/2/lib/two_bucket.dart
typedef Result = ({int moves, String goalBucket, int otherBucket});

class TwoBucket {

  int bucketOne = 0;
  int bucketTwo = 0;
  int goal = 0;
  String startBucket = '';
  
  TwoBucket({required this.bucketOne, required this.bucketTwo, required this.goal, required this.startBucket});
  
  List<int> gcdx(int a, int b){
    var res = List<int>.filled(3,0);
    if(b==0){
      res[0] = a; res[1] = 1; res[2] = 0;
    }
    else{
      res = gcdx(b, a%b);
      var u1 = res[2];
      res[2] = res[1] - res[2]*(a~/b);
      res[1] = u1;
    }
    return res;      
  }

  Result measureOne(int capBucket1, int capBucket2, int targetVol){
    if (targetVol == capBucket1)
      return (moves: 1, goalBucket: "one", otherBucket: 0);
    else if (targetVol == capBucket2)
      return (moves:2, goalBucket:"two", otherBucket:capBucket1);
    var gcd = gcdx(capBucket1, capBucket2);
    if(targetVol % gcd[0] !=0 )
      throw ArgumentError('impossible goal');
    int u = gcd[1] * (targetVol ~/ gcd[0]);
    int v = gcd[2] * (targetVol ~/ gcd[0]);
    if(u<0){
       int t = -u*gcd[0]~/capBucket2 + 1;
        u += (capBucket2~/gcd[0])*t;
        v -= (capBucket1~/gcd[0])*t;
    }
    if(capBucket1 > targetVol)
      return (moves: 2*(u-v)-2, goalBucket: "one", otherBucket: capBucket2);
    else
      return (moves: 2*(u-v), goalBucket: "two", otherBucket: 0);    
  }

  Result measure(){
    if(goal == 0)
      return (moves: 0, goalBucket: "one", otherBucket: 0);
    else if(goal > bucketOne
            && goal > bucketTwo)
      throw ArgumentError('goal larger than both buckets is impossible');
    if (startBucket == "one")
      return this.measureOne(bucketOne, bucketTwo, goal);
    else {
      Result res = this.measureOne(bucketTwo, bucketOne, goal);
      if (res.goalBucket == 'one') 
        return (moves: res.moves, goalBucket: 'two', otherBucket: res.otherBucket);
      else 
        return (moves: res.moves, goalBucket: 'one', otherBucket: res.otherBucket);
    }
  }
  
}
